rtfs.go

```diff
@@ -13,81 +13,70 @@ import (
 
 // IpfsManager is our helper wrapper for IPFS
 type IpfsManager struct {
-	Shell           *ipfsapi.Shell
-	PubSub          *ipfsapi.PubSubSubscription
-	KeystoreManager *KeystoreManager
-	KeystoreEnabled bool
-	PubTopic        string
-	nodeAPIAddr     string
+	PubTopic string
+
+	shell       *ipfsapi.Shell
+	keystore    *KeystoreManager
+	pubsub      *ipfsapi.PubSubSubscription
+	nodeAPIAddr string
 }
 
-// Initialize is used ot initialize our Ipfs manager struct
-func Initialize(pubTopic, connectionURL string) (*IpfsManager, error) {
-	manager := IpfsManager{
-		Shell:       EstablishShellWithNode(connectionURL),
-		PubTopic:    pubTopic,
-		nodeAPIAddr: connectionURL,
+// NewManager is used to initialize our Ipfs manager struct
+func NewManager(pubTopic, ipfsURL string, keystore *KeystoreManager) (*IpfsManager, error) {
+	// set up shell
+	sh := newShell(ipfsURL)
+	sh.SetTimeout(time.Minute * 5)
+	if _, err := sh.ID(); err != nil {
+		return nil, fmt.Errorf("failed to connect to ipfs node at '%s': %s", ipfsURL, err.Error())
 	}
-	manager.SetTimeout(time.Minute * 10)
-	_, err := manager.Shell.ID()
-	return &manager, err
-}
 
-// EstablishShellWithNode is used to establish our api shell for the ipfs node
-func EstablishShellWithNode(url string) *ipfsapi.Shell {
-	if url == "" {
-		shell := ipfsapi.NewShell("localhost:5001")
-		return shell
+	// subscribe to pubsub
+	sub, err := sh.PubSubSubscribe(pubTopic)
+	if err != nil {
+		return nil, fmt.Errorf("failed to subscribe to pub topic '%s': %s", pubTopic, err.Error())
 	}
-	shell := ipfsapi.NewShell(url)
-	return shell
+
+	// instantiate manager
+	return &IpfsManager{
+		PubTopic:    pubTopic,
+		shell:       sh,
+		pubsub:      sub,
+		nodeAPIAddr: ipfsURL,
+		keystore:    keystore,
+	}, nil
 }
 
-// SetTimeout is used to set a timeout for our api client
-func (im *IpfsManager) SetTimeout(time time.Duration) {
-	im.Shell.SetTimeout(time)
+// Close shuts down the manager
+func (im *IpfsManager) Close() error {
+	return im.pubsub.Cancel()
 }
 
-// CreateKeystoreManager is used to create a key store manager for ipfs keys
-// for now it just uses a file system keystore manager
-func (im *IpfsManager) CreateKeystoreManager() error {
```

Comment:
Looks like this was accidentally removed?

Reply:
Hmm nevermind it appears that it was an issue with the diff?

Reply:
this is intentional - i think its unidiomatic to have a constructor return a "partially instantiated" object, which it does in this case (with an empty keystore manager). It is now provided here:

https://github.com/RTradeLtd/rtfs/blob/cleanup/rtfs.go#L22

This allows you to provide whatever keystore you want, or not provide one at all

Reply:
Very good change :ok_hand: This will be extremely helpful for the eventual move to a secret management system like vault :fire: